Text/CustomEntitySearch/CustomEntitySearch.cs

```diff
@@ -52,48 +53,273 @@ public static class CustomEntitySearch
+                             caseSensitive = userInput.CaseSensitive;
+                         }
+
+                         var entities = new List<Entity>();
+                         var entitiesFound = new HashSet<string>();
+                         if (!string.IsNullOrWhiteSpace(text))
+                         {
+                             foreach (string word in words)
+                             {
+                                 if (string.IsNullOrEmpty(word)) continue;
+                                 int leniency = (exactMatches != null && exactMatches.Contains(word)) ? 0 : offset;
+                                 AddValues(word, text, entities, entitiesFound, leniency, caseSensitive);
+                                 if (synonyms != null && synonyms.ContainsKey(word))
+                                 {
+                                     foreach (string synonym in synonyms[word])
+                                     {
+                                         leniency = (exactMatches != null && exactMatches.Contains(synonym)) ? 0 : offset;
+                                         AddValues(synonym, text, entities, entitiesFound, leniency, caseSensitive);
+                                     }
+                                 }
+                             }
+                         }
+
+                         outRecord.Data["Entities"] = entities;
+                         outRecord.Data["EntitiesFound"] = entitiesFound;
+                     }
+                    return outRecord;
+                });
+
+            return new OkObjectResult(response);
+        }
+
+        public static void AddValues(string checkMatch, string text, List<Entity> entities, HashSet<string> entitiesFound, int leniency, bool caseSensitive)
+        {
+            bool addWord = false;
+            string escapedWord = Regex.Escape(checkMatch);
+            string pattern = (caseSensitive) ? @"\b(?x:" + escapedWord + @")\b" : @"\b(?ix:" + escapedWord + @")\b";
+            MatchCollection entityMatch = Regex.Matches(text, pattern, RegexOptions.Compiled, TimeSpan.FromSeconds(MaxRegexEvalTime));
+            if (entityMatch.Count != 0)
+            {
+                foreach (Match match in entityMatch)
+                {
+                    entities.Add(
+                        new Entity
+                        {
+                            Category = "customEntity",
+                            Value = match.Value,
+                            Offset = match.Index,
+                            Confidence = 1
+                        });
+                }
+                addWord = true;
+            }
+
+            if (leniency > 0)
+            {
+                // using KMP for now
+                // Create Table
+                IList<int> KMPTable = (caseSensitive) ? CreateKMPTable(checkMatch) : CreateKMPTable(checkMatch.ToLower());
+
+                // Begin searching!
+                int currTextChar = 0;
```

Comment:
>currTextChar [](start = 20, length = 12)

nit: this looks like an index not a char, consider renaming "currTextCharIndex"

same for word char below